File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def amountOfTime(self, root: Optional[TreeNode], start: int) -> int:
        graph = defaultdict(list)
        def dfs(node):
            if not node:
                return
            if node.left:
                graph[node.val].append(node.left.val)
                graph[node.left.val].append(node.val)
            if node.right:
                graph[node.val].append(node.right.val)
                graph[node.right.val].append(node.val)
            dfs(node.left)
            dfs(node.right)

        dfs(root)

        visited = set()
        q = deque([start])
        minutes = -1

        while q:
            minutes += 1
            for _ in range(len(q)):
                curr = q.popleft()
                visited.add(curr)
                for neib in graph[curr]:
                    if neib not in visited:
                        q.append(neib)

        return minutes

File: 2461-amount-of-time-for-binary-tree-to-be-infected/amount-of-time-for-binary-tree-to-be-infected.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
 *   var value: Int = _value
 *   var left: TreeNode = _left
 *   var right: TreeNode = _right
 * }
 */

import scala.collection.mutable.Map
import scala.collection.mutable.Set

object Solution {
    def amountOfTime(root: TreeNode, start: Int): Int = {
        
        val graph = Map.empty[Int, List[Int]].withDefaultValue(List[Int]())
        def dfs(root: TreeNode): Unit = {
            if (root != null) {
                if (root.left != null) {
                    graph(root.value) ::= root.left.value
                    graph(root.left.value) ::= root.value
                    dfs(root.left)
                }
                if (root.right != null) {
                    graph(root.value) ::= root.right.value
                    graph(root.right.value) ::= root.value
                    dfs(root.right)
                }
            }
            
            
        }

        dfs(root)

        val visited = Set(start)
        var minutes = -1
        var q = List(start)
        while (q.nonEmpty) {
            q = q.flatMap(graph(_)).filter(visited.add(_))
            minutes += 1
        }

        minutes
    }
}